### PR TITLE
Handle 64-bit systems and large memory allocations

### PR DESCRIPTION
**Pull request by manodeep**
Now can handle memory allocations greater than 2 GB. I have both a normal and a general solution - might be good to make a decision if the general solution is worth keeping around.

### Description

This pull request is to address a memory allocation issue identified in #8946

### CODE REVIEW

**Comment by manodeep** [comment]
@bsipocz In case there is any confusion -- the issue #8946 is different and might (will?) not be impacted by this PR

**Comment by bsipocz** [comment]
Ohh, OK, I didn't check the issue but rather misread your comment in the description.

**Comment by bsipocz** [comment]
I'm in the middle to wrap up the 4.0.1 release, so you can add the changelog entry once the release is out and the changelog is cleaned up (probably tomorrow).

File: astropy/utils/xml/src/iterparse.c [hunk 1]
```diff
@@ -40,13 +40,56 @@
...
+       
+       Limitations: Since a signed size_t (ssize_t) was required for the underlying CPython implementation,
+       on 32 bit systems, it will not be possible to allocate memory sizes between
+       SSIZE_MAX and SIZE_MAX (i.e, for allocations in the range [2^31, 2^32 - 1] bytes)
+       even though such memory might be available and accessible on the (32-bit) computer. That
+       said, since the underlying CPython implementation *also* uses Py_ssize_t (i.e., ssize_t),
+       it is safe to assume that such memory allocations would probably not be usable anyway. 
+       
+       TLDR: Will work on 64-bit machines but be careful on 32 bit machines when reading in 
+       ~ 2+ GB of memory -- ManodeepS 27/03/2020
```
Comment (pllim) [inline astropy/utils/xml/src/iterparse.c]: Maybe a more universal date formatting and GitHub username would be easier to read in the future.
```suggestion
       ~ 2+ GB of memory -- @manodeep 2020-03-27
```

File: astropy/utils/xml/src/iterparse.c [hunk 2]
```diff
@@ -40,13 +40,56 @@
...
     n |= n >> 4;
     n |= n >> 8;
     n |= n >> 16;
+    if(sizeof(Py_ssize_t) > 4) {
+        n |= n >> 32; /* this works for 64-bit systems but will need to be updated if (Py)_ssize_t 
+                         ever increases beyond 64-bits */
+    }
+#else    
+    /* This is the general solution for arbitrary widths for Py_ssize_t but is much less 
+    readable and possibly over-kill */
```
Comment (pllim) [inline astropy/utils/xml/src/iterparse.c]: Definitely harder to read but very appealing since we don't need to worry about `(Py)_ssize_t` increasing or not. I'll defer to C-ext experts.

File: astropy/utils/xml/src/iterparse.c [hunk 3]
```diff
@@ -40,13 +40,56 @@
...
+
+    /* If negative numbers are also allowed, then we need this line: 
+    if(n <= 1) {
+        return 1;
+    } 
+    */
+    
     n--;
+
+#if 1
```
Comment (pllim) [inline astropy/utils/xml/src/iterparse.c]: Note to maintainer: I think we need to pick one and remove this "if else".

File: astropy/utils/xml/src/iterparse.c [hunk 4]
```diff
@@ -40,13 +40,56 @@
...
+    if(sizeof(Py_ssize_t) > 4) {
+        n |= n >> 32; /* this works for 64-bit systems but will need to be updated if (Py)_ssize_t 
+                         ever increases beyond 64-bits */
+    }
+#else    
+    /* This is the general solution for arbitrary widths for Py_ssize_t but is much less 
+    readable and possibly over-kill */
+    const size_t numbits = sizeof(Py_ssize_t)*CHAR_BIT - 1;/* remove one for the sign-bit */
+    size_t shift = 1, totbits = 0;
+    while(totbits < numbits) {
```
Comment (mhvk) [inline astropy/utils/xml/src/iterparse.c]: How about the following to replace the whole routine? (Also takes care of `n <= 1`)
```
size_t result = 1;
while (n > 1) {
    result <<= 1;
    n >>= 1;
}
return result;
```

Comment (manodeep) [reply astropy/utils/xml/src/iterparse.c]: This is essentially what I used as the correct code in my [testing routine](https://gist.github.com/manodeep/110a39b43cb6472c108bc1fe13560819). The biggest difference is that the (worst-case) number of steps in this code snippet scales as the number of bits, while the other two solutions have a constant complexity of `O(log_2(nbits))`

Comment (mhvk) [reply astropy/utils/xml/src/iterparse.c]: Ah, I now see I really didn't look carefully enough at the original code! Suggestion withdrawn!

File: astropy/utils/xml/src/iterparse.c [hunk 5]
```diff
@@ -40,13 +40,56 @@
...
     n--;
+
+#if 1
+    /* This section just extends the existing solution to 64 bit systems */
     n |= n >> 1;
     n |= n >> 2;
     n |= n >> 4;
     n |= n >> 8;
     n |= n >> 16;
+    if(sizeof(Py_ssize_t) > 4) {
```
Comment (mhvk) [inline astropy/utils/xml/src/iterparse.c]: More of a question that anything else: this would be optimized away by the compiler, right? Since presumably it is fixed on any given system anyway. Might it be an idea regardless to make it a pre-compile `#if ...` so that that is obvious? With that, I think I'd stick with this simpler solution instead of the loop.

**Comment by pllim** [comment]
Why don't we wrap this up then?

1. Apply https://github.com/astropy/astropy/pull/10076#discussion_r400155433 and remove the loop implementation. (No need to squash commits.)
2. Add a change log under bug fix for 4.0.2.

Thank you!

File: astropy/utils/xml/src/iterparse.c (hunk 5 above)
Comment (manodeep) [reply astropy/utils/xml/src/iterparse.c]: Yeah - I think that `if` condition will (should?) get eliminated by all reasonable compilers under moderate optimisation. 

However, rewriting that `if` condition as a compile-time condition is a deceptively complicated task! `sizeof` calls are evaluated by the compiler and not the pre-processor. I looked into other possible ways of detecting 64-bit memory-addressing (you can have 64-bit addressing in a 32-bit system but I don't know what value `Py_ssize_t` takes in that case), but all the ones I found were not fully portable or restricted to specific compilers (`gcc` macros), common OS'es etc. 

The one solution I can think of is using:

```C
// at the top of the file
#ifndef PY_SSIZE_T_MAX
#define PY_SSIZE_T_MAX (SSIZE_MAX)
#endif

//Check if maximum addressable is > 2^31-1 (==0x7FFFFFFF)
#if PY_SSIZE_T_MAX > 0x7FFFFFFF
    n |= n >> 32;
#endif
```

But this solution depends on `SSIZE_MAX` being defined. While `SSIZE_MAX` is defined on all POSIX systems, I am unsure what happens on Windows.

Basically, I am not confident in the pre-processor solution and it might break the build. I ma happy to give this solution a shot, and revert the fix if the CI chokes.

Comment (mhvk) [reply astropy/utils/xml/src/iterparse.c]: That's curious.  I just checked numpy, and it uses a `__LP64__`:
https://github.com/numpy/numpy/blob/f0a74b2d4e50a1b4f9d9189c6b2e31b920913a8b/numpy/core/include/numpy/numpyconfig.h#L15-L22

That seems some kind of standard; see https://stackoverflow.com/questions/6721037/where-is-lp64-defined-for-default-builds-of-c-applications-on-osx-10-6

Comment (manodeep) [reply astropy/utils/xml/src/iterparse.c]: That's specifically for OSX. You will need similar definitions on all supported OS'es + compilers. 

For most systems, the presence of any of the macros `__x86_64__`, `__LP64__`, `__amd64__`, `__IA64__` indicates that 64-bit compilation is in effect. On Windows, `_M_X64`, or `_WIN64` indicates 64-bit compilation. On ARM64 systems, the GNU C standard defines `__aarch64__` (think that requires compiling with `-std=gnu*`). Unsure what happens on Cray, Solaris, other "unusual" OS'es. [Source for pre-defined macros on various systems](https://sourceforge.net/p/predef/wiki/Architectures/).

@mhvk I am happy to code that in using `__LP64__` if you think that will work within the astropy-supported OS'es. However, in trying to figure out how to properly code this at the pre-processor stage, I am somewhat certain that the current implementation with the `if` condition is the *most* robust.

Comment (mhvk) [reply astropy/utils/xml/src/iterparse.c]: Duh, I should really have looked more carefully. The numpy macro undefines and then redefines it for apple, but I cannot even find where it is defined in the first place. In princple, I guess we could import the numpy macros and just use theirs. But I am starting to agree that we might as well let the compiler optimize it away! Or should one just unconditionally run the extra instruction? If the compiler is smart enough it will realize that one is or'ing with 0 and optimize that away as well...

Anyway, I feel I'm causing you far more work than it's worth! Just go ahead and do what you feel is best, and let's get this in!

Comment (manodeep) [reply astropy/utils/xml/src/iterparse.c]: It was a fun exercise for the pedant in me :D. Really did not expect implementing the `if` condition as a (portable) pre-processor condition to be this difficult!

Bit-shifting in excess of the bit-width of a variable (that would occur on 32-bit systems if we always execute the last right-shift by 32 bits) will be undefined behaviour and allows the compiler complete freedom to replace/remove/do anything the compiler wants. Think it's best to stick to the `if` condition.

Comment (mhvk) [reply astropy/utils/xml/src/iterparse.c]: OK, so let's go with the explicit solution. Could you remove the loop part?  I think it might also be good to rewrite the comment a bit: the solution is now good for either 32 or 64 bit, but on 32 bit cannot use the highest one.

**Comment by manodeep** [comment]
@pllim How should I pull in the latest changelog (before adding in the entry within `v4.0.2`)? 

Also, while we are editing this file, should I uncomment [the lines](https://github.com/manodeep/astropy/blob/patch-1/astropy/utils/xml/src/iterparse.c#L62-L66) that handle negative arguments? or simply remove those lines?

(First time-committer, please excuse the ignorance)

**Comment by pllim** [comment]
To grab latest changes from `master` (assuming you name this repo remote `upstream`):

```
git fetch upstream master
git rebase upstream/master
```

Unless you run into conflicts, the above commands should be all you need. You also need to force push afterwards. Make a backup branch first if in doubt. Thanks for working on this!

> negative arguments?

Technically, it is out of scope, right? This PR explicitly states that it addresses 64-bit problem, not negative value. So, I guess, unless you have strong opinion otherwise, to remove it.

**Comment by bsipocz** [comment]
@manodeep - feel free to ping me on slack if you run into any issues with the rebase

**Comment by manodeep** [comment]
@pllim Made the updates the way you suggested - think that went without any fuss. Please let me know if I need to make any other changes. Thanks!

**Comment by mhvk** [comment]
@manodeep - sorry for the hassle, but could you rebase again, but this time also squash all the commits into one? I don't think we need to preserve the history here. Thanks!

**Comment by manodeep** [comment]
@mhvk Ohh not a problem at all - I was wondering that during the rebase. I totally agree - these individual commits do not add any value to the history. Let me give that a go

**Comment by manodeep** [comment]
@mhvk Updated with a rebase + squash. Seems like it went fine and everything got squashed into one commit.

**Comment by bsipocz** [comment]
Thanks @manodeep! And I'm looking forward for future contributions!
